Transition to rust-analyzer as our official LSP (Language Server Protocol) implementation (#2912)
rust-lang/rfcs: text/0000-rust-analyzer.md

- Feature Name: (fill me in with a unique ident, `my_awesome_feature`)
- Start Date: 2020-04-20
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
- master [compiler team IDE strategy and planning](/9qrJYdSLTK2MWG4-pbmpWw?edit) document

# Summary
[summary]: #summary

The RFC proposes a plan to adopt rust-analyzer as Rust's official LSP implementation. The transition to rust-analyzer will take place in a few stages:

* **Feedback** -- encourage people to use rust-analyzer and report problems

> * **Deprecation** -- actively transition people from the RLS to rust-analyzer
> * **Transition** -- stop supporting the older RLS

Reviewer: Deprecation and Transition seem to mixed up, it should look like this:
```suggestion
* **Transition** -- actively transition people from the RLS to rust-analyzer
* **Deprecation** -- stop supporting the older RLS
```

> As detailed below, one major concern with rust-analyzer as it stands today is that it shares very little code with rustc. To avoid creating an unsustainable maintainance burden, this RFC proposes extracting shared libraries that will be used by both rustc and rust-analyzer ("library-ification"), which should eventually lead to rustc and rust-analyzer being two front-ends over a shared codebase.

Reviewer 2: Please not “library-ification”. “librarification” or “librification”.

Reviewer 3: While the original wording does seem a little silly  I perosnally find it way easier to read and understand than the suggested contracted versions.

Reviewer 3: While the original wording does seem a little silly  I perosnally find it way easier to read and understand than the suggested contracted versions.
EDIT: Also, English in general is not really a language where you join words together like that, maybe "Library Extraction" serves the purpose better?

Reviewer 4: Library-fication would be most [exact](https://www.yourdictionary.com/fication), if they intend to describe the goal.  
If they are doing it now its ying/y-ification.

Reviewer 4: @kiljacken the whole is becoming a collection of libraries, not only a part

Reviewer 5: Not sure how much we're invested in using the actual term but maybe _modularization_ would be better? (It does remind me of Rust modules, though...)

Reviewer 4: Since Rust reads very nicely like a book, I prefer the book-related term. Arization comes from arize comes from arise which sounds like something pops up unexpected. ;-)

Author: lol, I am for some odd reason quite fond of the "obviously incorrect" library-ification spelling, and it's definitely what we say more commonly when speaking out loud and typing informally, in contrast to librarification. 

I don't like "modularization", even though I agree it is more correct, because it doesn't really communicate the "scale" to me.  (As you said, @Xanewok, it sounds like Rust modules.)

In any case, I.. don't really want to debate this particular terminology right here, as it is hardly central to the RFC. Maybe we can come up with some clever pun based on the term *crates* though...

In contrast, rust-analyzer effectively reimplements the Rust compiler in a fully incremental, on-demand style. This is the same architecture that rustc has been slowly evolving towards. This architecture enables much faster response time and it can also (in principle) handle things like fully type-correct completions. However, because rust-analyzer is not complete, it is currently not able to offer several key features, such as reporting errors or doing precise "find all usages".

Even in its current, experimental form, many users derive value from rust-analyzer. Many users are using it as their day-to-day IDE. It is particularly useful for larger codebases, such as the compiler.

## Challenges to overcome

There are several things that we would like to improve about the current situation:

> * We would to concentrate our efforts behind one LSP server, not support both the RLS and rust-analyzer.

Reviewer 6: ```suggestion
* We would like to concentrate our efforts behind one LSP server, not support both the RLS and rust-analyzer.
```

    * Further, the goal for some time has been to adopt a query-based architecture much like the one that rust-analyzer is using.
* We would like to (eventually) avoid having two implementations of the Rust compiler to support, one in rustc and one in rust-analyzer.
* We would like to "pay down" technical debt within the compiler itself and to make it approachable.
    * To that end, we've been pursuing the creation of independent libraries, like miri or chalk. Smaller libraries with stronger API boundaries are easier to reason about but also provide an easier way for people to get involved in compiler development.

However, in making the transition from the existing RLS setup to rust-analyzer, we have to be careful not to introduce user confusion. In particular, we wish to make the experience of "managing one's editor" smooth, both for:

> * existing RLS users (who need to transition from the RLS to rust-analyzer), and
> * new Rust useres (who need to find and install rust-analyzer for the first time).

Reviewer 7: nit: Inconsistent capitalization and typo

```suggestion
* Existing RLS users (who need to transition from the RLS to rust-analyzer), and
* New Rust users (who need to find and install rust-analyzer for the first time).
```

## Separate goal: making the compiler more approachable via 'library-ification'

> Independently from IDEs, The compiler team has been pursuing a process of "library-ification", meaning convering rustc from a monolithic codebase into one with well-defined libraries and reasonably stable API boundaries. You can find more details in the [design meeting from 2019-09-13][2019-09-13]. The goal is ultimately for both rustc and rust-analyzer to be shallow wrappers around the same core codebases, as well as to improve the accessibility of the rustc codebase by having well-defined modules that can be learned independently.

Reviewer 6: ```suggestion
Independently from IDEs, The compiler team has been pursuing a process of "library-ification", meaning converting rustc from a monolithic codebase into one with well-defined libraries and reasonably stable API boundaries. You can find more details in the [design meeting from 2019-09-13][2019-09-13]. The goal is ultimately for both rustc and rust-analyzer to be shallow wrappers around the same core codebases, as well as to improve the accessibility of the rustc codebase by having well-defined modules that can be learned independently.
```

## Observation: the needs of batch compilation and the needs of an IDE are not always the same

One observation that we have seen over time is that batch compilation and IDE interaction have somewhat different needs. We would like to share as much code as possible, but we might like to specialize some aspects of it.

For example, the compiler currently interns all of its types and frees them all at once at the end of compilation. This is highly efficient but not necessarily appropriate for a long-lived process, as over time it can lead to very high memory usage. An IDE, in contrast, might prefer to use a different strategy such as ref-counting or even garbage collection.

Similarly, in rustc, we have been moving towards a model where the dependency graph between queries is streamed out to the disk as soon as it is generated, and never stored in memory. This is because the dependency information is only needed when you start the *next* compilation. But in an IDE, that dependency information is needed as soon as the next keypress, and hence it doesn't make sense to stream it to disk.

> Library-ification can address these concerns by two distinct "host processes" that make use of shared libraries differently. In the case of types, for example, we can be generic over whether types ar interned or stored in some other sort of pointer. Similarly, the query infrastructure might have two modes or implementations strategies. 

Reviewer 6: ```suggestion
Library-ification can address these concerns by two distinct "host processes" that make use of shared libraries differently. In the case of types, for example, we can be generic over whether types are interned or stored in some other sort of pointer. Similarly, the query infrastructure might have two modes or implementation strategies. 
```

[r-a-install]: https://rust-analyzer.github.io/manual.html#installation

The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

### Timeline for transition

Transition will occur in three phases:

* **Feedback:** During this first phase, we will post a blog post encouraging RLS users to try out rust-analyzer and see whether it works for them. If we encounter unexpected, blocking issues, or cases where people feel rust-analyzer is a significant step backward in their user experience, we may try to fix those issues before fully replacing the existing RLS.
* **Deprecation period:** We announce that support for the RLS is deprecated. We begin putting in place the tooling to transition existing users away from the RLS.
* **Final transition:** We no longer support the RLS plugin in its older form and no longer distribute RLS over rustup.

### What is the transition plan?

> The precise transition plan is not part of this RFC. It will be determined and announced as we enter the Deprecation period, based on the feedback we've gotten and how many users have manually transitioned away from the RLS. We will endeavor to keep the experience as smooth as possible, but it may require some manual steps.

Reviewer 6: ```suggestion
The precise transition plan is not part of this RFC. It will be determined and announced as we enter the deprecation period, based on the feedback we've gotten and how many users have manually transitioned away from the RLS. We will endeavor to keep the experience as smooth as possible, but it may require some manual steps.
```

### Branding: how to talk about rust-analyzer/RLS going forward?

* We propose to keep the "rust-analyzer" name, at least for the transition period.
* In keeping with the [proposed rust-lang github access policy](https://github.com/rust-lang/rfcs/pull/2872), the repositories from the [rust-analyzer github org](https://github.com/rust-analyzer) will be consolidated and then merged into the [rust-lang github org](https://github.com/rust-lang).
    * They will be maintained by the compiler team with a dedicated working group.

>     * The infra team will work with rust-analyzer to integrate the binary release and upgrade process

Reviewer 8: If the intent is to integrate `rust-analyzer` into the main release channels and handle it through `rustup` then at least some work will be needed in `rustup` (e.g. provision of proxy)

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

The primary alternative considered was to halt work on rust-analyzer and instead attempt to port the lessons from its development to rustc. In effect, the idea would be to create a LSP server based on rustc itself.

The primary appeal of this plan is that there would always be a single codebase. Moreover, the fundamental architecture of rustc has been moving steadily towards the demand-driven, IDE-friendly design that rust-analyzer has also adopted (the two have indeed influenced one another), so this would be a natural extension of that work.

However, there are a number of practical concerns with taking that approach. One concern is that, for rust-analyzer's current users, it would represent a regression. rust-analyzer would no longer be available (or at least no longer updated) and it would take some time until rustc is at feature parity. Moreover, experience has shown that refactoring rustc can move relatively slowly, simply due to the age of the codebase, the amount of code involved, and the complicated non-standard build process.

> Further, the 'reimplement' approach would represent a "constraint" on the ordering in which we do our work. With the design proposed in this RFC, for example, rust-analyzer is able to make use of the chalk library already. This is only possible because rust-analyzer has a "stub" version of Rust's name resolution engine and type checker embedded in it -- this type checker is not perfect, but it's good enough to drive chalk and gain useful experience. This allows us to create an end-to-end IDE user experience sooner, in effect.

Reviewer 6: Nit: Is the mixing of single quotes and double quotes intended?

# Prior art
[prior-art]: #prior-art

The current proposal is informed by experience with existing RLS and query-based compilation in rustc. Additionally, rust-analyzer heavily draws from lessons learned while developing IntelliJ Rust.

It's interesting that many compilers went through a phase with parallel implementaitons to get a great IDE support

* For C#, the [Roslyn](https://github.com/dotnet/roslyn) project was a from scratch implemenation.
* [Dart for a long time had different front-ends for command line and interractive compilers](https://youtu.be/WjdrUphF5l4?t=2204)
* [Swift is transitioning to new syntax tree library by "reimplement separately, then swap" approach](https://medium.com/@kitasuke/deep-dive-into-integrating-libsyntax-into-the-compiler-pipeline-2d478c8600a1)

Notable exceptions:

* Kotlin, TypeScript -- these languages were implemented with IDEs in mind from the start

> * OCaml with merlin and C++ with clangd -- languages with header files and forward declarations make it easier to adapt traditional compiler architecture to certain IDE tasks like completion.

Reviewer 9: Note, that the OCaml had to rethink it's merlin and LSP architecture, thus the [ocaml-lsp](https://github.com/ocaml/ocaml-lsp) project. As it turned out, the original Merlin architecture wasn't suitable enough for the VSCode and Vim/Emacs LSP clients.

- Feature Name: (fill me in with a unique ident, `my_awesome_feature`)
- Start Date: 2020-04-20
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
- master [compiler team IDE strategy and planning](/9qrJYdSLTK2MWG4-pbmpWw?edit) document

# Summary
[summary]: #summary

The RFC proposes a plan to adopt rust-analyzer as Rust's official LSP implementation. The transition to rust-analyzer will take place in a few stages:

* **Feedback** -- encourage people to use rust-analyzer and report problems
* **Deprecation** -- actively transition people from the RLS to rust-analyzer
* **Transition** -- stop supporting the older RLS

> As detailed below, one major concern with rust-analyzer as it stands today is that it shares very little code with rustc. To avoid creating an unsustainable maintainance burden, this RFC extracting shared libraries that will be used by both rustc and rust-analyzer ("library-ification"), which should eventually lead to rustc and rust-analyzer being two front-ends over a shared codebase.

Reviewer 10: ```suggestion
As detailed below, one major concern with rust-analyzer as it stands today is that it shares very little code with rustc. To avoid creating an unsustainable maintainance burden, this RFC proposes extracting shared libraries that will be used by both rustc and rust-analyzer ("library-ification"), which should eventually lead to rustc and rust-analyzer being two front-ends over a shared codebase.
```

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

The high-level plan is effectively to adopt rust-analyzer as the primary LSP implementation for the Rust project, and to aggressively pursue 'library-ification' as a means to eliminate code duplication. The ultimate vision is that the majority of the compiler logic should live in shared libaries which have two "front-ends", one from rustc and one from rust-analyzer.

## Adopting rust-analyzer as the primary LSP implementation

### Installing rust-analyzer today

Today, [to install rust-analyzer for VSCode][r-a-install], one simply installs the rust-analyzer plugin. The plugin will download the rust-analyzer LSP implementation and automatically keep it up to date (currently on a once-per-week release cadence). The plugin also adds the `rust-src` component to rustup.

[r-a-install]: https://rust-analyzer.github.io/manual.html#installation

> The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

Reviewer 11: This is the same situation that the RLS is in, although with more established plugins. Given the LSP architecture, I don't think things will get worse with Rust Analyzer rather than RLS

> The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

Reviewer 10: @nrc That's fair, but we did _not_ have an official statement (to my knowledge) that we only support one editor. I think that's important: either we should say nothing or we should support most common options.

> The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

Reviewer 12: Looking at https://blog.rust-lang.org/2020/04/17/Rust-survey-2019.html , there seems to be a clear break between the popularity of VSCode/Vim/Intellij and the other options. Further, I think it is important to look at the trajectory of popularity -- I doubt Emacs or Sublime will suddenly become much more widely used than they are now.

Additionally, the "IDE setups" question appears to show that most IntelliJ users sensibly use the IntelliJ Rust plugin, and neither RLS nor rust-analyzer.

From this data, I agree that Vim should have good rust-analyzer support, but focusing energy now on support for editors beyond VSCode and Vim seems a bit premature. Of course, a high level of support for many editors is a great long-term goal.

> The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

Reviewer 13: We moved the rust-analyzer integration for Emacs to the emacs-lsp project because it makes more sense for it to be maintained there. Vim is in a similar situation: There are multiple LSP plugins, and they typically maintain their support for specific language servers. For VSCode on the other hand, there isn't one central plugin for the language server integration, rather each language server has to maintain their own plugin.

> The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

Reviewer 14: This is a good question, to which, unfortunately, I don't know a good answer. Let me give some context which explains the current state.

*First*, my thinking is that, in the grand scheme of things, particular editors, or even the particular LSP protocol, are not that important. The hard bit is the incremental lazy compiler, changing its "front-ends" should be relatively (emphasis on relatively) painless. So internally rust-analyzer speaks in Rust-specific terms, which are dumbed down to LSP lowest common denominator at the boundary. Additionally, rust-analyzer goes beyond standard LSP, to make sure that important IDE features are implementable. For example, we shipped extend selection and syntax highlighting well before they were accepted into the protocol proper (and experience in rust-analyzer informed the protocol design process).

*Second*, I want to emphasize that the current de facto situation with VS Code is not the result of someone's editor preferences. I haven't used VS Code much when starting rust-analyzer, and I researched a bunch of options. I settled on VS Code (despite my preexisting bias towards Emacs) because it is meaningfully easier to write a language server plugin for it. Although today some editors have caught up a bit, I think this conclusion still stands today. 

That's it for background. Here are some statements about the current situation.

*First* I believe there's unfortunately a technical reason why supporting VS Code is much simpler:

* it has a de-facto single LSP implementation
* the implementation is a library with a stable interface, which you can use to implement separate language specific plugins

This is in contrast to other editors, which in generally have many competing implementations, and which have a "monorepo" with all language plugins. We used to have Emacs support in tree, but it was moved to emacs-lsp repository.

*Second*, I think though, in theory, LSP should work with any editor, in practice, support for it in VS Code is meaningfully better. Like, the fact that the best plugin for Vim for rust-analyzer runs nodejs to be able to use Microsoft's LSP library exemplifies the situation. "making rust-analyzer working in $EDITOR$ as good as it is in VS Code" in practice involves hacking, first and foremost, on the $EDITOR$ itself. 

*Third*, it is somewhat impossible to not have a blessed implementation. People who hack on rust-analyzer most will naturally fix their editor's clients for them. Additionally, when implementing outside-of-the-lsp features (like the hypothetical lifeitmes visualizer), they need to be prototyped in some client. 

*Fourth*, not everyone has their favorite editor. I think for a lot of users (especially for new users) having a blessed solution is beneficial, because choosing between many editors is hard, if you don't care about the editor and just want to learn the language. 

*Fifth*, it's instructive to take a look at the editor usage of [Go](https://blog.golang.org/survey2019-results#TOC_5.), where IntelliJ (Goland) + VS Code account for 75% and are clearly growing.

> The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

Reviewer 14: With all this in mind, I think a reasonable strategy is:

* continue to officially develop vs code plugin
* be ready to officially support other editors, when 1) the editors have an official production-ready LSP implementation 2) there are people on rust-analyzer team willing to put effort into this
* in the official docs, provide links to editor specific quick-starts (for which the VS Code one will be maintained by the rust project, and the others will be maintained by community).

> The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

Author: It seems like part of @mark-i-m's concern is the "officialness and permanence" of having VSCode as the "main" plugin. My assumption was that this text was meant more to describe the current state of affairs than to declare a permanent policy. I think it's pretty clear that right now the attention of the team should be focused on the "guts" of rust-analyzer, and that -- for this purpose -- having a single main plugin makes sense, to avoid undue distraction. We should definitely link to other editor plugins and help users to get started, of course. But I imagine that once things settle down it's entirely possible that we would maintain, or at least contribute heavily to, support for other editors.

> The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

Reviewer 5: I'd like to chime in and propose that for VSCode specifically we continue to use the existing [`rust-lang.rust`](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust) plugin and implement support for the Rust Analyzer engine.

This is the most popular downloaded extension for Rust with almost 0.5M downloads now and should considerably reduce the churn and smooth the transition period, especially if it's going to be as easy as prompting the user with a "Using Rust Analyzer engine is preferred now. Switch?" modal.

It's something I've been meaning to do for a while now and this sounds like a perfect opportunity, along with rebranding of the extension from "RLS client" to officialy supported Rust plugin for VSCode (especially since it's published by the "official" [rust-lang](https://marketplace.visualstudio.com/publishers/rust-lang) which hopefully adds to the visibility and credibility).

> The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

Reviewer 15: > the vim support is rather a pain to get working (though once it does it works great), and requires having a nodejs server running

There are multiple client options and I assume you're talking about the coc-rust-analyzer route, what I can say is that for me there's simple no pain, it's trivial to install/setup, and second, having the nodejs server for this particular client is actually good, since the rationale for its existence is to be close to what vscode is able to provide so that it's easy to fork vscode extensions for working in [neo]vim and migrate features. If one doesn't like this trade off, then they should use other solutions. Just wanted to mention that because people keep complaining about it having nodejs while never mentioning why it's there.

> The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

Reviewer 10: Hi @matklad. Thanks for your response and rationale and for all your hard work on ra!

I think that strategy makes sense, but IMHO, the phrasing of the RFC should clearly indicate that this choice of editors is purely for alpha-level development and that it is our intent to support other editors as it become practical.

> Fifth, it's instructive to take a look at the editor usage of Go, where IntelliJ (Goland) + VS Code account for 75% and are clearly growing.

I think there are 2 reasons why this is not a good example for us to follow:

1.  Personally, I haven't used go in a few years, but the last time I used it, the debugging experience was pretty not-great with gdb and other standard tools. It's my (unverified) impression that goland/vscode tend to get adopted because they work better than the alternatives, not because people prefer those platforms. I am concerned that looking at numbers like the one you cited will result in the following vicious cycle: most people use vscode => we de-emphasize other editors => experience with other editors remains bad => people throw up their hands and use vscode => most people use vscode. So in such a world most people would use vscode, but not really by choice. And I have a strong impression that a lot of the adoption of vscode in the rust community is due to this sort of reaction.

2. It might make sense to have special tools for languages like go, java, js, etc, where there is an extensive runtime. But for rust, I expect to be able to use the same tools as I use for C/C++ development. That includes standard terminal utilities and debuggers: readelf, objdump, ld, ar, gdb, etc, and an arbitrary editor.

> The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

Reviewer 16: > Personally, I haven't used go in a few years, but the last time I used it, the debugging experience was pretty not-great with gdb and other standard tools. It's my (unverified) impression that goland/vscode tend to get adopted because they work better than the alternatives, not because people prefer those platforms.

FWIW, [the 2019 Go Developer Survey also showed significant dissatisfaction with debugging](https://blog.golang.org/survey2019-results), although their analysis didn't consider any possible correlations with IDE usage.

-----

And yes, the current wording of the RFC is giving _many_ readers a completely unintended impression of the long-term plan here (I had the same knee-jerk concern myself), so we definitely need to rephrase it.

> The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

Reviewer 17: > With all this in mind, I think a reasonable strategy is:
> 
> * continue to officially develop vs code plugin
> * be ready to officially support other editors, when 1) the editors have an official production-ready LSP implementation 2) there are people on rust-analyzer team willing to put effort into this

On the second of these points it might be worth mentioning that [Neovim](https://neovim.io/) has added [a language server client to the editor](https://github.com/neovim/neovim/pull/11336) itself, although I don't think it's in a stable release yet. This should help provide a more stable LSP interface in this slice of the vim eco-system. Also, if history is anything to go by it seems likely the regular vim will follow suit.

Anyway, that is to say the situation may improve in the future and for now, "continue to officially develop vs code plugin", seems reasonable.

> The experience of installing rust-analyzer for other editors is more varied. The rust-analyzer project only directly supports VSCode.  Other editor plugins are maintained independently.

Author: I added a sentence that we may to add official support for other editors later.
